src/base/hide_password.cpp

```diff
@@ -0,0 +1,861 @@
+  * CREATE SERVER srv1 (HOST='localhost', PORT=3300, DBNAME=demodb, USER=dev1)
+  *
+  * CREATE USER user_name
+  * CREATE USER user_name PASSWORD 'password_string'
+  * CREATE USER user_name PASSWORD _utf8'password_string'
+  * CREATE USER user_name PASSWORD _euckr'password_string'
+  * ALTER USER user_name PASSWORD 'password_string'
+  * ALTER SERVER srv1 CHANGE PASSWORD='password string'
+  */
+
+  *password_len = 0;
+  *is_found_pwd_keyword = false;
+  while (*ps)
+    {
+      token = get_token (ps, len);
+      if (*token == ';' || *token == ')')
+	{
+	  return token;
+	}
+
+      if (len == 8 && strncasecmp (token, "password", 8) == 0)
+	{
+	  *is_found_pwd_keyword = true;
+	  break;
+	}
+    }
+
+  if (*ps == '\0')
+    {
+      // check case "CREATE USER user_name"
+      return (is_create && !is_server) ? ps : NULL;
+    }
+
+  prev = NULL;
+  token = get_token (ps, len);
+  if (is_server)
+    {
+      /* pattern cases)
+         *    PASSWORD = 'string'   PASSWORD = ''    PASSWORD = )
+         *    PASSWORD = ,          PASSWORD = ;     PASSWORD = \0
+        */
+      if (*token != '=')
+	{
+	  return NULL;
+	}
+
+      token = get_token (ps, len);
+      if (*token == ';' || *token == ')' || *token == ',' )
+	{
+	  return token;
+	}
+    }
+  else if (*token == '_')
+    {
+      prev = token;
+      tlen = len;
+      if ((len == 5 && strncasecmp (token + 1, "utf8'", len) == 0)
+	  || (len == 6 && strncasecmp (token + 1, "euckr'", len) == 0)
+	  || (len == 9 && strncasecmp (token + 1, "iso88591'", len) == 0)
+	  || (len == 7 && strncasecmp (token + 1, "binary'", len) == 0))
+	{
+	  token = get_token (ps, len);
+	}
+      else
+	{
+	  return NULL;
+	}
+    }
+
+  if (len >= 2 && (*token == '\'' || *token == '"'))
+    {
+      *password_len = (prev ? (len + tlen) : len);
+      return (prev ? prev : token);
+    }
+
+  return NULL;
+}
+
+char *
+hide_password::skip_one_query (char *query)
+{
+  char *ps;
+
+  for (ps = query; *ps; ps++)
+    {
+      if (*ps == '\'' || *ps == '"')
+	{
+	  ps = get_quot_string (ps);
+	  if (*ps == '\0')
+	    {
+	      break;
+	    }
+	}
+      else if (*ps == '-')
+	{
+	  if (ps[1] == '-')
+	    {
+	      ps = skip_comment_string (ps);
+	    }
+	}
+      else if (*ps == '/')
+	{
+	  if (ps[1] == '/' || ps[1] == '*')
+	    {
+	      ps = skip_comment_string (ps);
+	    }
+	}
+      else if (*ps == ';')
+	{
+	  ps++;
+	  break;
+	}
+    }
+
+  return ps;
+}
+
+bool hide_password::check_lead_string_in_query (char **query, char **method_name, bool *is_create, bool *is_server)
+{
+  char   *ps;
+  const char *first_cmd_str[] = { "call", "create", "alter", NULL };
+  int  first_cmd_len[] = { 4, 6, 5, -1 };
+  const char *second_cmd_str[] = { "server", "user", NULL };
+  int  second_cmd_len[] = { 6, 4, -1 };
+  static const char *method_name_str[] = { "add_user", "set_password", "login", NULL };
+  int method_name_len[] = { 8, 12, 5, -1 };
```

Comment:
trivial, please ignore:
int method_name_len[] = { 
  strlen (method_name_str[0]),
  strlen (method_name_str[1]),
  strlen (method_name_str[2]),
-1 };

Reply:
I'll keep the current code for this part.
However, I will reinforce some of the comments.